### PR TITLE
Fix CLI tests

### PR DESCRIPTION
**Pull request by Tryneus**
This is a bit of a refactor of the CLI because literally everything was broken by all the branches that just managed to land at about the same time.  The primary goal is to get the tests working, but I'm also addressing many inconsistencies across files in the cli module.

It's not quite done yet, but most of the hard work should be there.  As it stands, the scripts and tests were doing some pretty horrifying things, like returning early (and thus doing stuff concurrently), leaving rethinkdb running or connections open, using old functions that no longer exist, etc, etc.

Anyway, it's 7 AM, so I'll resume this tomorrow, but putting it up now in case anyone wants to make comments earlier on.

---

This change is [Reviewable](https://reviewable.io/reviews/rethinkdb/horizon/696)

### CODE REVIEW

**Comment by Tryneus** [comment]
Oh, the biggest problem I noticed was that there was a circular dependency between `schema.js` and `serve.js`, which meant that the `serve` object was empty inside `schema` - meaning nothing should have ever worked - the tests were just so broken it didn't show up.

File: cli/test/schema.spec.js [hunk 1]
```diff
@@ -31,90 +33,83 @@ indexes = ["datetime"]
...
-      });
-    });
-  });
+  let rdb_server;
+  let rdb_conn;
+
+  before('start rethinkdb', () =>
+    start_rdb_server({
+      bind: ["127.0.0.1"],
+      dataDir: tmpdir() + "/horizon-test",
```
Comment (dalanmiller) [inline cli/test/schema.spec.js]: One thing I forgot to add here was we should save this generated path and then remove it after the tests are done or have exited. 

Maybe this?:

``` js
describe("hz schema", () => {
  let rdb_server, rdb_conn;
  const tmpPath = tmpdir() + "/horizon-test"; 

  before('start rethinkdb', () => 
    start_rdb_server({
      bind: ["127.0.0.1"],
      dataDir: tmpPath,
    }).then((server) => {
      rdb_server = server;
    });
  );

  after('stop rethinkdb', () => {
    // Delete temporary database dir
    fs.rmdirSync(tmpPath);
    // Close RDB server connection
    return rdb_server.close();
  });

  ...
```

File: cli/test/schema.spec.js [hunk 2]
```diff
@@ -31,90 +33,83 @@ indexes = ["datetime"]
...
-    it("saves schema to schema.toml from rdb", () => {
-      return runSaveCommand({
+    it("saves schema to schema.toml from rdb", () =>
+      runSaveCommand({
         start_rethinkdb: false,
         rdb_host: '127.0.0.1',
         rdb_port: currentOpts.dbPort,
-        out_file: fs.createWriteStream('out.toml', {flags: 'w'}),
+        out_file: fs.createWriteStream('out.toml', { flags: 'w' }),
         project_name: 'horizon_schema_test',
```
Comment (dalanmiller) [inline cli/test/schema.spec.js]: This line should just be `project_name` using the `const` defined at the top

File: cli/test/schema.spec.js [hunk 3]
```diff
@@ -125,66 +120,39 @@ describe("hz schema", () => {
...
```
Comment (dalanmiller) [inline cli/test/schema.spec.js]: This should also probably use the `project_name` `const`.

**Comment by dalanmiller** [comment]
@Tryneus you must be Midas, because all these tests have been turned to gold.

File: cli/test/schema.spec.js [hunk 4]
```diff
@@ -31,90 +33,83 @@ indexes = ["datetime"]
...
-    
-    beforeEach('set up database state', () => {
-      // Create MockFS
-      mockFs({
-          '.hz': {
-            'schema.toml': testSchema,
-          },
-          'out.toml': '',
-      });
+    //beforeEach('initialize mockfs', () => {
```
Comment (dalanmiller) [inline cli/test/schema.spec.js]: Should this be commented out?

**Comment by Tryneus** [comment]
All the tests should be more or less passing at this point.  Some of the cli tests can be rather slow as they involve creating several RethinkDB tables in the setup phase.

File: cli/test/schema.spec.js (hunk 3 above)
Comment (Tryneus) [inline cli/test/schema.spec.js]: +1, fixed in the latest upload

File: cli/test/schema.spec.js (hunk 2 above)
Comment (Tryneus) [inline cli/test/schema.spec.js]: fixed in the latest upload

File: cli/test/schema.spec.js (hunk 1 above)
Comment (Tryneus) [inline cli/test/schema.spec.js]: I added a stage to clean up the directory after the suite runs.

File: cli/test/schema.spec.js (hunk 4 above)
Comment (Tryneus) [inline cli/test/schema.spec.js]: Ah, I wasn't completely sure if I'd be breaking anything by removing this as I'd added a mockFs initialization higher up in the suite.  Turns out this was no longer necessary, so I removed it.

File: cli/src/utils/initialize_joi.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+'use strict';
+
+// Issues a dummy joi validation to force joi to initialize its scripts.
+// This is used because tests will mock the filesystem, and the lazy
+// `require`s done by joi will no longer work at that point.
+module.exports = (joi) =>
+  joi.validate('', joi.any().when('', { is: '', then: joi.any() }));
```
Comment (dalanmiller) [inline cli/src/utils/initialize_joi.js:7]: According to @marsup, calling `Joi.options()` would also be sufficient.

https://gitter.im/hapijs/joi?at=57a1236747659bfb108977f7

Comment (Marsup) [inline cli/src/utils/initialize_joi.js:7]: Just checked, `Joi.options({})` actually.

**Comment by dalanmiller** [comment]
A nitpick: it seems that I can no longer use `console.log` in the test files to output some debugging variables as the logging is being squelched somewhere. 

I also couldn't even do `process.stdout.write('something')` and have it appear in the output.

File: cli/src/utils/initialize_joi.js (hunk 5 above)
Comment (dalanmiller) [inline cli/src/utils/initialize_joi.js:7]: Thanks @Marsup! 👍

File: cli/src/serve.js [hunk 6]
```diff
@@ -132,7 +127,7 @@ function parseArguments(args) {
 
   parser.addArgument([ '--config' ],
     { type: 'string', metavar: 'PATH',
-      help: 'Path to the config file to use, defaults to "${default_config_file}".' });
+      help: 'Path to the config file to use, defaults to "${config.default_config_file}".' });
 
   parser.addArgument([ '--schema-file' ],
     { type: 'string', metavar: 'SCHEMA_FILE_PATH',
```
Comment (dalanmiller) [inline cli/src/serve.js:133]: Extra `"` in help for `--schema-file`

**Comment by deontologician** [comment]
:lgtm: mergin'

**Pull request merged**